WIP: vir ast copying and Expr enum refactoring

Maxwell-Yang-2001 opened the pull request:
Copied files from [prusti-common/src/vir/ast](https://github.com/viperproject/prusti-dev/tree/master/prusti-common/src/vir/ast) to under vir crate in a newly created [polymorphic directory](https://github.com/Maxwell-Yang-2001/prusti-dev/tree/polymorphic-vir/vir/defs/polymorphic). Removed any impl or defined traits, but still had links to prusti-common as of now for [here](https://github.com/Maxwell-Yang-2001/prusti-dev/blob/polymorphic-vir/vir/defs/polymorphic/expr.rs#L8) and [here](https://github.com/Maxwell-Yang-2001/prusti-dev/blob/polymorphic-vir/vir/defs/polymorphic/stmt.rs#L8) (which before moving was using contents under the same crate), since only part of ast in vir is copied.

Refactored the enum `Expr` in [`expr.rs`](https://github.com/Maxwell-Yang-2001/prusti-dev/blob/polymorphic-vir/vir/defs/polymorphic/expr.rs) into several structs, currently located in the [same file](https://github.com/Maxwell-Yang-2001/prusti-dev/blob/polymorphic-vir/vir/defs/polymorphic/expr.rs#L95-L239) (could be moved to a directory structure if needed). Naming of struct fields are based on [comments](https://github.com/viperproject/prusti-dev/blob/master/prusti-common/src/vir/ast/expr.rs#L17-L56) and [usages on this enum](https://github.com/viperproject/prusti-dev/blob/master/prusti-common/src/vir/ast/expr.rs#L103-L219).

@vakaras thanks in advance!

vakaras committed:
Fix after_expiry parsing.

Aurel300 committed:
implement existential quantifiers

dario23 committed:
Add encode_place parameter for whether to encode shared or mutable array indexing

dario23 committed:
Implement mutable array access

dario23 committed:
Cleanup

dario23 committed:
Add mutable array referencing testcase

JustinHuPrime committed:
Wrote buggy recursive descent preparser.

JustinHuPrime committed:
Add disabled prusti_expr! test files

JustinHuPrime committed:
Noted parser's accepted EBNF

dario23 committed:
Implement in-place array mutation

dario23 committed:
Add lookup_pure triggers

dario23 committed:
Remove useless `clone()` calls

JustinHuPrime committed:
fix bad use of pop

Maxwell-Yang-2001 committed:
copied files from ast and refactored expr

copy borrows to polymorphic and update old links

update stmt to use structs

update polymorphic structure and add legacy

fix macro issue and remove vir in prusti-common

Revert "fix macro issue and remove vir in prusti-common"

This reverts commit 39a5e10fa294145458883a9b360b373ae68568c5.

Revert "update polymorphic structure and add legacy"

This reverts commit 8274623433faaa2d1b7c9e828aa94d2863a424eb.

remove dependency from prusti-common

vakaras commented on vir/Cargo.toml:
```diff
@@ -10,9 +10,11 @@ serde = { version = "1.0", features = ["derive"] }
 syn = { version = "1.0", features = ["full", "fold", "parsing", "derive"] }
 quote = "1.0"
 proc-macro2 = { version = "1.0", features = ["span-locations"] }
+prusti-common = { path = "../prusti-common" }
```
We need to get rid of this dependency because `prusti-common` will depend on `vir`.

vakaras commented on vir/defs/polymorphic/stmt.rs:
```diff
@@ -0,0 +1,78 @@
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+use super::*;
+use prusti_common::vir::borrows::{Borrow, DAG as ReborrowingDAG};
```
Please also copy `borrows` so that we can remove dependency on `prusti_common`.

vakaras replied:
Let's put it inside `vir/src`. Then, the following import should work:

```rust
use crate::borrows::{Borrow, DAG as ReborrowingDAG};
```

Maxwell-Yang-2001 replied:
And this `borrows.rs` also relies on part of `polymorphic`. By being placed outside of `polymorphic` (so without being generated as part of `vir.rs`), how would it import things from `vir.rs`? I think the main confusion for me is how `vir.rs` would be located relative to other files or directories in the `vir` crate.

vakaras replied:
Oh, I missed that `borrows.rs` depends on VIR. In that case, it needs to be inside `defs/polymorphic`.

vakaras replied:
> I think the main confusion for me is how vir.rs would be located relative to other files or directories in the vir crate.

The generated `vir.rs` is put into the `target` directory. It is included by the include macro in `vir/src/lib.rs`, so you can think as all code from `vir.rs` is inlined at that place.

JustinHuPrime committed:
add purpose comments

JustinHuPrime committed:
get rid of paren-nesting handling

paren nesting is already handled here

JustinHuPrime committed:
add special case for empty assertions

JustinHuPrime committed:
fix handling of groups

JustinHuPrime committed:
wrote down pledge grammar

dario23 committed:
Encode index place in expand_one_level

dario23 committed:
Exhale arrays like local variables in loop permissions

We need to find the maximum used permission for each array, so they are
first collected and then emitted at the end.

dario23 committed:
Add two more testcases

dario23 committed:
Revert to manual fmt::Debug impl for Node

This one has less info, but the derived one is too noisy for everything
else than debugging.

dario23 committed:
Properly set position for exhale statements, re-enable assert

dario23 committed:
Set pos for array ref expiration stmts as well

dario23 committed:
Avoid separate loop variable in tests

dario23 committed:
Introduce ExprOrArrayBase enum for PlaceEncoding

dario23 committed:
Document and simplify ProcedureEncoder.array_magic_wand_at

JustinHuPrime committed:
fix parsing of parenthesized parts of rust expressions

JustinHuPrime committed:
fix parsing of entailments with more than one assertion

vakaras commented on prusti-common/src/vir/mod.rs:
```diff
@@ -4,21 +4,7 @@
-mod conversions;
-pub mod fixes;
-pub mod optimizations;
-mod to_viper;
```
I think that it is better to move only the AST/CFG nodes and leave all other code here. For example, I think modules `to_viper`, `optimizations`, and `fixes` should stay.

vakaras commented on vir/Cargo.toml:
```diff
@@ -10,9 +10,16 @@ serde = { version = "1.0", features = ["derive"] }
+log = { version = "0.4", features = ["release_max_level_info"] }
+config = "0.11"
+itertools = "0.10.0"
+viper = { path = "../viper" }
```
`vir` should not depend on `viper`.

vakaras commented on vir/build.rs:
```diff
@@ -13,6 +13,11 @@ fn main() {
     let root = quote! { components =>
+        pub mod legacy;
+        pub mod polymorphic;
+        pub mod config;
```
`config` should not be here because it has nothing to do with VIR.

vakaras commented on vir/build.rs:
```diff
@@ -13,6 +13,11 @@ fn main() {
+        pub mod legacy;
+        pub mod polymorphic;
+        pub mod config;
+        pub mod utils;
```
`utils` probably also can be left in their original place for now.

vakaras commented on vir/build.rs:
```diff
@@ -13,6 +13,11 @@ fn main() {
+        pub mod polymorphic;
+        pub mod config;
+        pub mod utils;
+        pub mod report;
```
`report` should also be left in the original place.

dario23 committed:
Update error message for non-supported array RHS cases

JustinHuPrime committed:
somewhat more accurate error message spans

dario23 committed:
Add a testcase for the complex array RHS error

dario23 committed:
Match Slices in PlaceEncoding

dario23 committed:
Add Seq, ContainerOp to vir::Expr

dario23 committed:
Implement array snapshotting

dario23 committed:
Move encode_array_types, encode_slice_types to separate array_encoder

This way we avoid making the encoder.rs file even longer

dario23 committed:
Move encoding of lookup_pure to Encoded{Array,Slice}Types struct

dario23 committed:
Allow pre-encoding sequence element snapshot type

This is useful when calling `encode_array_types` from e.g. inside the
`SnapshotEncoder` where `Encoder`'s `snapshot_encoder` field is already
borrowed, so `encoder.encode_snapshot_type` would result in a double
borrow (i.e. panic).

dario23 committed:
Add caching to array_types_encoder

dario23 committed:
Use array snapshot encoding in pure_function_encoder

dario23 committed:
Add simple array snapshot testcase

JustinHuPrime committed:
pledge parsing

dario23 committed:
Handle preserving array contents in loops

They can't be as easily copied as scalar values, so we use snapshots
instead.

dario23 committed:
Do not take val_fields of arrays

dario23 committed:
Add testcase for passing array to pure function

dario23 committed:
Change predicate syntax to allow implications

Constructs like Prusti's implication `==>` that aren't valid Rust were
not possible in previous style `#[predicate]` functions, because rustc
parsed the functions already for e.g. name resolution, before we were
able to rewrite them. With the new function-like proc-macro aproach, we
avoid that problem.

dario23 committed:
Implement enough of array slicing to call .len() on arrays

dario23 committed:
Update syn to at least 1.0.43

impl syn::Parse for syn::Signature was added in 1.0.43, and we're using
it in parsing predicate function signatures.

Maxwell-Yang-2001 committed:
clean up legacy

ani003 committed:
Merging fix for panics in encode_copy2

JustinHuPrime committed:
tweaked error messages

Aurel300 committed:
Merge pull request #539 from ani003/master

Fixes for panics in encode_copy2

JustinHuPrime committed:
tweak error messages and bless new outputs

dario23 committed:
Address reviews

dario23 committed:
Add docstrings for EncodedArrayTypes, EncodedSliceTypes

ani003 committed:
Changed unimplemented to unsupported error

dario23 committed:
Don't encode type in lookup_pure name

it'll be added later anyway as part of the argument types, so we'll just
go with "lookup_pure" before mangling

ani003 committed:
Update error message

Co-authored-by: Aurel <Aurel300@users.noreply.github.com>

Aurel300 committed:
Merge pull request #532 from dario23/array-types-encoding

Refactor array types encoding

ani003 committed:
Added test for encode_const_expr

fpoli committed:
Add from_nth example of pledge

dario23 committed:
Simplify hashmap insert

dario23 committed:
Add comment why we use mir_encoder's encode_place

Aurel300 committed:
Merge pull request #541 from ani003/panic_fix

Changed unimplemented to unsupported error

dario23 committed:
Remove resolved TODO

dario23 committed:
Use Expr::map_old_expr_label

dario23 committed:
misc fixes from review

dario23 committed:
Use (manual) snapshotting for idx.val_int encoding

Aurel300 committed:
Merge pull request #489 from dario23/array-idx-mut

Encoding of mutable array indexing

Aurel300 committed:
reenable some passing tests from tests_old

fpoli committed:
Merge pull request #542 from viperproject/from-nth

Add from_nth example of pledge

ani003 committed:
Replaced unreachable panic with generic error message

ani003 committed:
Added test for region of projection panic

ani003 committed:
Fix: Custom error message for each variant

fpoli committed:
Fix parse/ui test annotations

fpoli committed:
Fix typecheck/ui test annotations

fpoli committed:
Fix verify/ui test annotations

dario23 committed:
Use lookup_pure for specs

actual pure functions will have their arrays be snapshots, so we do
read(snap, idx) there, but interacting with what the surrounding
procedure knows in specification code is easier when just encoding as
lookup_pure(a, idx).

dario23 committed:
Add successfully verifying selection sort testcase

fpoli committed:
Merge pull request #530 from viperproject/update-rustc

Update rustc version

dario23 committed:
Try to better specify which unsize casts to allow

dario23 committed:
hackily ignore unsupported raw_ref invariant

the invariants (that are always empty) are generated on the raw
variable, but other preconditions are generated on .val_ref, so we run
into consistency errors otherwise.

Aurel300 committed:
Merge pull request #544 from Aurel300/reenable-tests

Reenable some passing tests from `tests_old`

dario23 committed:
Return unsupported error for iterator methods early

They would encode to viper without crashing just fine, but then have
subtle other problems, so just avoid them for now.

fpoli committed:
Fix computation of nonconditional loop blocks

dario23 committed:
Update testcase with iterators message

fpoli committed:
Enable test

fpoli committed:
Merge pull request #546 from viperproject/fix-loop-checks

Fix computation of nonconditional loop blocks

dario23 committed:
Rework labeling for array equalities in loops

dario23 committed:
Remove superfluous space

Co-authored-by: Aurel <Aurel300@users.noreply.github.com>

Aurel300 committed:
handle argumentless attributes when unwrapping tokens

dario23 committed:
Move elem snapshot type out of EncodedArrayTypes

dario23 committed:
Don't store element snapshot type for slices either

Aurel300 committed:
Merge pull request #216 from viperproject/trait-refinement

Fix after_expiry parsing.

fpoli committed:
Add extensible test suite for cargo-prusti

fpoli committed:
Remove debug statement

fpoli committed:
Fix name of test crate

fpoli committed:
Set the "prusti" feature when running Prusti

fpoli committed:
Add expected stderr of cargo_verify/failing_crate

JustinHuPrime committed:
don't construct intermediate pledge

fpoli committed:
Add Cargo.lock to keep cargo_verify tests reproducible

JustinHuPrime committed:
update tests with stdout files
also update ignored tests with new proposed syntax

fpoli committed:
Remove warnings, fix nondeterminism in SLF4J

fpoli committed:
Compile test crates before testing cargo-prusti

fpoli committed:
Fix nondeterminism on MacOS

fpoli committed:
Fix paths on Windows

dario23 committed:
Don't create false snapshot types in snapshot array encoding

dario23 committed:
Recursively format sequence type when computing function identifier

fpoli committed:
Merge pull request #547 from viperproject/test-cargo-prusti

Add extensible test suite for cargo-prusti

dario23 committed:
Remove unnecessary space

Co-authored-by: Aurel <Aurel300@users.noreply.github.com>

dario23 committed:
Expr::Seq fallible_fold without temporary variables

it's a fold, after all :-)

dario23 committed:
Use longer `array_types` variable name consistently

dario23 committed:
Use DomainFunc::apply instead of manual Expr::domain_func_app

dario23 committed:
Simplify match

ani003 committed:
Changed return type to EncodingResult

Return type of encode_value_expr and encode_value_field in encoder.rs changed to EncodingResult to propagate the errors returned by type_encoder.encode_value_field

dario23 committed:
Don't explicitly snap arrays when indexing

JustinHuPrime committed:
applied small code review items

JustinHuPrime committed:
avoid cloning when possible

fpoli committed:
Remove debug statements

fpoli committed:
Merge pull request #552 from viperproject/test-cargo-prusti

Remove debug statements

fpoli committed:
Migrate to GitHub Viper releases

fpoli committed:
Fix removal of viper_tools

fpoli committed:
Remove SLF4J output

fpoli committed:
Fix find_viper_home

fpoli committed:
Fix Prusti setup on Windows

dario23 committed:
Add space around field index in projection error

JustinHuPrime committed:
Merge branch 'master' into refactor-parser

fpoli committed:
Update rustc to nightly-2021-06-09

Aurel300 committed:
Update prusti-interface/src/environment/mir_utils/split_aggregate_assignment.rs

Aurel300 committed:
Merge pull request #538 from dario23/array-slice-len

Implement enough of array slicing to call .len() on arrays

Aurel300 committed:
Merge pull request #554 from dario23/projection-msg

Add space around field index in projection error

Aurel300 committed:
Merge pull request #551 from ani003/return_result

Returning result type from encode_value_expr and encode_value_field

Aurel300 committed:
fix merge error

ani003 committed:
Exhausted pattern match

Exhausted the pattern match and returning unsupported error from the newly added patterns

fpoli committed:
Merge pull request #553 from viperproject/viper-releases

Migrate to GitHub Viper releases

dario23 committed:
Shorten footprint collection for vir::Expr::Seq

dario23 committed:
Emit EncodingError::unsupported instead of panicking

dario23 committed:
Fix indent

Co-authored-by: Aurel <Aurel300@users.noreply.github.com>

dario23 committed:
Remove unneeded comment

dario23 committed:
Predicate has no more attributes, so remove test function

ani003 committed:
Added unsupported error

vakaras committed:
Fix deleting the viper_tools directory.

vakaras committed:
Clarify the error message.

ani003 committed:
Return unsupported error instead of panic

Aurel300 committed:
improve snapshots of argumentless enums

ani003 committed:
Streamlining code

Aurel300 committed:
hacky fix for issue with Silicon

Aurel300 committed:
Merge pull request #561 from Aurel300/fix/snapshot-enums

Improve snapshots of argumentless enums

Aurel300 committed:
Merge pull request #537 from dario23/predicates

Change predicate syntax to allow nested implications

ani003 committed:
Added testcase

Testcase for unsupported higher-ranked lifetimes and types error

dario23 committed:
Add equality to lookup_pure as array snap postcondition

dario23 committed:
Also patch snapshots in DomainFuncApp

pretty similar to how FuncApp is handled

vakaras committed:
Update dependencies.

dario23 committed:
Arrays support equality as well

dario23 committed:
Use snapshots for everything in pure function encoder

dario23 committed:
add two more testcases

dario23 committed:
Assume that snapshot read is the same as lookup

For big arrays prusti/viper had problems verifying, so we'll just go
with assuming it.

dario23 committed:
Implement slices in pure functions and specifications

dario23 committed:
Add testcase for slices in pure functions and specifications

Aurel300 committed:
Merge pull request #505 from dario23/array-snap-spec

Array snapshot encoding and specs

dario23 committed:
Trigger on lookup_pure as well in lookup_pure equality

dario23 committed:
Use a.len() for non-specification code

For specification code, we need more slice support in the pure function
encoder.

dario23 committed:
Register bounds check ErrorCtxt for slices as well

dario23 committed:
Add pure function slice out-of-bounds testcase

dario23 committed:
Length request never happens on anything other than arrays or slices

dario23 committed:
Also trigger on lookup call for slice read axiom

dario23 committed:
Rework conditions, len encoding

dario23 committed:
Encode snapshotting operation for unsize-cast in pure code

dario23 committed:
Add two testcases for slices in pure code

Aurel300 committed:
Merge pull request #566 from dario23/verify-selectionsort

Verify selectionsort

ani003 committed:
removed assert; return error from general case

Removed the assert as it is covered in the pattern match below. Returning an error in the final case instead of unimplemented panic

dario23 committed:
Verify mergesort's merge function on static arrays

Slices aren't far enough along yet, but the main (difficult) loop of
mergesort can be verified for two static pre-sorted sub-arrays already

Aurel300 committed:
Merge pull request #560 from ani003/add_error_type_vis

Add unsupported error function to TypeVisitor

Aurel300 committed:
Merge pull request #558 from ani003/type_enc_fix

Exhausted pattern match

dario23 committed:
Update prusti-tests/tests/verify/pass/arrays/merge.rs

Co-authored-by: Aurel <Aurel300@users.noreply.github.com>

ani003 committed:
Removed unnecessary returns

Aurel300 committed:
Merge pull request #568 from dario23/verify-mergesort

Verify mergesort's merge function on static arrays

ani003 committed:
Fix compilation error

dario23 committed:
Add slice split_at testcase

This wraps the built-in slice split_at method, to show we have all the
tools needed to ad specifications for that now.

We can't properly use #[extern_spec], because that doesn't work with
`impl [T]` for slices yet.

Aurel300 committed:
Merge pull request #545 from ani003/panic_fix_regions

Replaced unreachable panic with error

ani003 committed:
Added testcase

Testcase corresponding to PR number 558

dario23 committed:
Add assertion about splitted slice contents

Aurel300 committed:
Merge pull request #572 from ani003/add_test

Added testcase

ani003 committed:
Fix panic

ani003 committed:
Exhausted pattern match; returning errors from new cases

ani003 committed:
Added testcase

Raw pointers unsupported test case

dario23 committed:
use vir::Function::apply

shorter and less error-prone

dario23 committed:
implement array snapshot slicing

dario23 committed:
Add slice_helper for slices as well

dario23 committed:
Add another testcase

Aurel300 committed:
Merge pull request #580 from ani003/mir_enc_fix

Remove unnecessary assert

Aurel300 committed:
Merge pull request #577 from ani003/typaram_fix

Fixing panic in typaram

Aurel300 committed:
Merge pull request #578 from ani003/proc_enc_1320

Exhausting pattern match

Aurel300 committed:
implement constant index patterns

Aurel300 committed:
Merge pull request #582 from Aurel300/feature/constantindex-pattern

Implement `ConstantIndex` patterns

ani003 committed:
Added function to convert Tykind to String

ani003 committed:
Replaced panics with internal error

ani003 committed:
Fix warning

ani003 committed:
Added Unsupported variant to FoldUnfoldError

ani003 committed:
Return unsupported error for two stage borrow

ani003 committed:
Added test for two stage borrow

ani003 committed:
Changed error message

ani003 committed:
Streamlining code

Co-authored-by: Aurel <Aurel300@users.noreply.github.com>

ani003 committed:
Modify error message

Modified error message based on review

Co-authored-by: Aurel <Aurel300@users.noreply.github.com>

JustinHuPrime committed:
Merge branch 'master' into refactor-parser

Aurel300 committed:
Merge pull request #549 from JustinHuPrime/refactor-parser

Refactor Parser into Pure Recursive Descent Parser

ani003 committed:
Fix panic caused by non-empty tuple constants

ani003 committed:
Delete code marked for deletion

ani003 committed:
Fixed unimplemented panic

vakaras committed:
Update Viper toolchain.

ani003 committed:
Return unsupported error for variadic functions

ani003 committed:
Added testcase for constant tuples

ani003 committed:
Fix github warning

Aurel300 committed:
Merge pull request #564 from dario23/slice-snap

Add snapshot encoding for slices

ani003 committed:
Added unsupported error for unions

ani003 committed:
Added testcase for unions

vakaras committed:
Only report verification success if there were no encoding errors.

vakaras committed:
Fix a bug in x.py setting of the overflow flag.

Aurel300 committed:
take a snapshot when encoding old(...)

Maxwell-Yang-2001 committed:
update traits for polymorphic VIR

Maxwell-Yang-2001 committed:
update traits for polymorphic VIR

Aurel300 committed:
fix snapshot patching in quantifiers, binops

Aurel300 committed:
patch snapshots in spec functions (will be replaced by closures PR anyway)

Aurel300 committed:
Merge pull request #602 from Aurel300/issue/599

Take a snapshot when encoding old(...)

fpoli committed:
Update rustc to nightly-2021-07-15

Aurel300 committed:
Merge pull request #585 from ani003/tykind_string

Tykind to string function

Aurel300 committed:
Merge pull request #587 from ani003/add_to_foldunfolderror

Adding Unsupported variant to FoldUnfoldError for two phase borrow

Aurel300 committed:
Merge pull request #592 from ani003/const_panic_fix

Fix panic caused by constant tuples

Aurel300 committed:
Merge pull request #593 from ani003/fix_real_edges

Fix unimplemented panic

Aurel300 committed:
Merge pull request #597 from ani003/union_panic_fix

Unsupported error for unions

ani003 committed:
Added testcase for variadic functions

Aurel300 committed:
Merge pull request #595 from ani003/variadic_err

Return unsupported error for variadic functions

vakaras committed:
Update dependencies.

vakaras committed:
Fix UI tests.

vakaras committed:
Fix UI tests.

vakaras committed:
Fix UI tests.

vakaras committed:
Fix a bug in preparser.

vakaras committed:
Add some tests for exists.

vakaras committed:
Rename ForAllVars to QuantifierVars.

vakaras committed:
Make prusti-contracts to compile on stable.

vakaras committed:
Update Rustc version to 2021-07-17.

vakaras committed:
Make Clippy a separate job.

vakaras committed:
Panic if a user specifies unexpected configuration flag.

vakaras committed:
Try to fix the test to work in release mode.

vakaras committed:
Add missing configuration variables.

Maxwell-Yang-2001 committed:
implement For for polymorphic->legacy (without substitution)

Maxwell-Yang-2001 committed:
implement For for polymorphic->legacy (without substitution)

Maxwell-Yang-2001 committed:
Merge branch 'polymorphic-vir' of https://github.com/Maxwell-Yang-2001/prusti-dev into polymorphic-vir

Maxwell-Yang-2001 committed:
update new Enum variants from master to vir
